lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs

```diff
@@ -17,49 +17,46 @@
 
 using System;
 using System.Collections.Generic;
+using Org.Apache.REEF.Common.Tasks;
 using Org.Apache.REEF.Common.Tasks.Events;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Interface;
+using Org.Apache.REEF.Utilities.Collections;
 
 namespace Org.Apache.REEF.Common.Runtime.Evaluator.Task
 {
     internal sealed class TaskLifeCycle
     {
-        private readonly HashSet<IObserver<ITaskStop>> _taskStopHandlers;
-        private readonly HashSet<IObserver<ITaskStart>> _taskStartHandlers;
-        private readonly ITaskStart _taskStart;
-        private readonly ITaskStop _taskStop;
+        private readonly IReadOnlyCollection<IObserver<ITaskStop>> _taskStopHandlers;
+        private readonly IReadOnlyCollection<IObserver<ITaskStart>> _taskStartHandlers;
+        private readonly IInjector _injector; 
 
-        // INJECT
-        public TaskLifeCycle(
-            HashSet<IObserver<ITaskStop>> taskStopHandlers,
-            HashSet<IObserver<ITaskStart>> taskStartHandlers,
-            TaskStartImpl taskStart,
-            TaskStopImpl taskStop)
+        [Inject]
+        private TaskLifeCycle(
+            [Parameter(typeof(TaskConfigurationOptions.StartHandlers))] ISet<IObserver<ITaskStart>> taskStartHandlers,
+            [Parameter(typeof(TaskConfigurationOptions.StopHandlers))] ISet<IObserver<ITaskStop>> taskStopHandlers,
+            IInjector injector)
         {
-            _taskStartHandlers = taskStartHandlers;
-            _taskStopHandlers = taskStopHandlers;
-            _taskStart = taskStart;
-            _taskStop = taskStop;
+            _taskStartHandlers = new ReadOnlySet<IObserver<ITaskStart>>(taskStartHandlers);
+            _taskStopHandlers = new ReadOnlySet<IObserver<ITaskStop>>(taskStopHandlers);
+            _injector = injector;
         }
 
-        public TaskLifeCycle()
-        {
-            _taskStartHandlers = new HashSet<IObserver<ITaskStart>>();
-            _taskStopHandlers = new HashSet<IObserver<ITaskStop>>();
-        }
-        
         public void Start() 
         {
-            foreach (IObserver<ITaskStart> startHandler in _taskStartHandlers) 
+            var taskStart = _injector.GetInstance<TaskStartImpl>();
+            foreach (var startHandler in _taskStartHandlers)
             {
-                startHandler.OnNext(_taskStart);
+                startHandler.OnNext(taskStart);
             }
         }
 
         public void Stop() 
         {
-            foreach (IObserver<ITaskStop> stopHandler in _taskStopHandlers)
+            var taskStop = _injector.GetInstance<TaskStopImpl>();
```

Comment:
Usually we inject from interface assume impl is set in configuration in the injector.